sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4003,38 +4046,84 @@ piEnqueueMemBufferMap(pi_queue Queue, pi_mem Buffer, pi_bool BlockingMap,
     ZeEvent = (*Event)->ZeEvent;
   }
 
+  // TODO: Level Zero is missing the memory "mapping" capabilities, so we are
+  // left to doing new memory allocation and a copy (read) on discrete devices.
+  // On integrated devices  we have allocated the buffer in host memory
+  // so no actions are needed here except for synchronizing on incoming events
+  // and doing a host-to-host copy if a host pointer had been supplied
+  // during buffer creation.
+  //
+  // TODO: for discrete, check if the input buffer is already allocated
+  // in shared memory and thus is accessible from the host as is.
+  // Can we get SYCL RT to predict/allocate in shared memory
+  // from the beginning?
+  if (BufferUsesHostMem) {
+    (*Event)->HostSyncforMap = true;
+#if 0
+    for (uint32_t i = 0; i < NumEventsInWaitList; i++) {
+      zePrint("Map added ZeWaitEvent = %lx\n",
+              pi_cast<std::uintptr_t>(EventWaitList[i]->ZeEvent));
+      (*Event)->waitEvents.push_back(EventWaitList[i]->ZeEvent);
+    }
+#else
+    for (uint32_t i = 0; i < NumEventsInWaitList; i++) {
```

Comment:
just call piEventsWait here

Reply:
OK. Also have to move the queue lock down.